orttraining/orttraining/python/training/utils/hooks/_statistics_subscriber.py

```diff
@@ -0,0 +1,111 @@
+# -------------------------------------------------------------------------
+# Copyright (c) Microsoft Corporation. All rights reserved.
+# Licensed under the MIT License.
+# --------------------------------------------------------------------------
+
+from pathlib import Path
+from typing import Union
+
+import os
+import warnings
+import shutil
+import torch
+
+from ._subscriber_base import SubscriberBase
+
+
+class StatisticsSubscriber(SubscriberBase):
+    """
+    This subscriber is used to dump the activation's statistics to files.
+
+    Each activation will be summarized into 1 or 2 files, depending on whether it is used in the backward pass.
+    So for each run step, there will be many files.
+    > In the forward pass, summarize the tensor's statistics and write to a file.
+    > In the backward pass, summarize the tensor's gradient's statistics and write into another file.
+
+    Currently the statistics mainly include:
+    > Number of inf/nan values.
+    > Common statistics: tensor shape, data type, total element size, min/max/mean/std of the tensor elements.
+    > Number of zero/negative/positive elements.
+    > Sampled elements (taking the first 128 elements for example).
+    > To be extended...
+
+    `merge_activation_summary.py` can be used to merge the files into one file per training step.
+    """
+
+    def __init__(
+        self,
+        output_dir: str,
+        start_step: Union[None, int] = None,
+        end_step: Union[None, int] = None,
+        override_output_dir: bool = False,
+    ):
+        """
+        Steps in [start_step, end_step) will run subscriber actions.
+        :param output_dir: the directory all activation statistics files will be stored.
+        :param start_step: the first step that runs subscriber actions.
+        :param end_step: the end step (exclusively) that runs subscriber actions.
+        :param override_output_dir: whether `output_dir` can be overridden if it already exists.
```

Comment:
If I remember correctly, we use google stype guide for python documentation: https://google.github.io/styleguide/pyguide.html.

Any reason for using restructuredtext?

Reply:
Good point

Reply:
Fixed along with other typos, gramma errors.